pkg/provider/azure_loadbalancer.go

```diff
@@ -371,10 +371,39 @@ func (az *Cloud) cleanOrphanedLoadBalancer(lb *network.LoadBalancer, service *v1
 
 		// Remove the LB.
 		klog.V(10).Infof("cleanOrphanedLoadBalancer(%s, %s, %s): az.DeleteLB starts", lbName, serviceName, clusterName)
-		err = az.DeleteLB(service, lbName)
-		if err != nil {
-			klog.V(2).Infof("cleanOrphanedLoadBalancer(%s, %s, %s): failed to DeleteLB: %v", lbName, serviceName, clusterName, err)
-			return err
+		deleteErr := az.DeleteLB(service, lbName)
+		if deleteErr != nil {
+			klog.V(2).Infof("cleanOrphanedLoadBalancer(%s, %s, %s): failed to DeleteLB: %v", lbName, serviceName, clusterName, deleteErr)
+
+			rgName, vmssName, parseErr := retry.GetVMSSMetadataByRawError(deleteErr.Error())
+			if parseErr != nil {
+				klog.Warningf("cleanOrphanedLoadBalancer(%s, %s, %s): failed to parse error: %v", lbName, serviceName, clusterName, parseErr)
+				return deleteErr
+			}
+			if rgName == "" || vmssName == "" {
+				klog.Warningf("cleanOrphanedLoadBalancer(%s, %s, %s): empty rgName or vmssName", lbName, serviceName, clusterName)
+				return deleteErr
+			}
+
+			// if we reach here, it means the VM couldn't be deleted because it is being referenced by a VMSS
+			if _, ok := az.VMSet.(*ScaleSet); !ok {
+				klog.Warningf("cleanOrphanedLoadBalancer(%s, %s, %s): unexpected VMSet type, expected VMSS", lbName, serviceName, clusterName)
+				return deleteErr
+			}
+
+			// build a ipConfigID and try to remove that reference first and retry delete lb
+			ipConfigIDs := []string{fmt.Sprintf("/subscriptions/sub/resourceGroups/%s/providers/Microsoft.Compute/virtualMachineScaleSets/%s/virtualMachines/vmss-0/networkInterfaces/0", rgName, vmssName)}
+			err := az.VMSet.(*ScaleSet).ensureBackendPoolDeletedFromVMSS(service, lbBackendPoolID, vmSetName, ipConfigIDs)
```

Comment:
this makes some assumptions for the format of ip config. Could we invoke `EnsureBackendPoolDeleted` here?

Reply:
I don't think we can call `EnsureBackendPoolDeleted` directly because there is no `lb.backendPool.BackendIPConfigurations` in this scenario. https://github.com/kubernetes-sigs/cloud-provider-azure/blob/bd9b6cdcfada69109353f16f3292ad2bc68388f7/pkg/provider/azure_vmss.go#L1533

Reply:
thanks for the refactoring. Could you check why basic lb CI is failing?